### PR TITLE
Add Rails support to Ruby module

### PR DESCRIPTION
**Pull request by alexandergraul**
This is WIP, I've picked the upstream keyindings with minor changes for now.

Fixes https://github.com/hlissner/doom-emacs/issues/2461, https://github.com/hlissner/doom-emacs/issues/2275

### CODE REVIEW

File: modules/lang/ruby/config.el [hunk 1]
```diff
@@ -163,3 +164,55 @@
...
+        :desc "Find Job" "b" #'projectile-rails-find-job
+        :desc "Extract Region" "x" #'projectile-rails-extract-region
+        ;; Go to files
+        :desc "Goto Gemfile" "gg" #'projectile-rails-goto-gemfile
+        :desc "Goto Routes" "gr" #'projectile-rails-goto-routes
+        :desc "Goto Schema" "gd" #'projectile-rails-goto-schema
+        :desc "Goto Seeds" "gs" #'projectile-rails-goto-seeds
+        :desc "Goto Spec Helper" "gh" #'projectile-rails-goto-spec-helper)
+        ;; Open consoles etc.
+        (:prefix ("o" . "open")
```
Comment (hlissner) [inline modules/lang/ruby/config.el]: Please remove these which-key labels. They do not work on minor mode keymaps (the keys are bound absolutely, meaning all `<localleader> r` keybinds will have these labels, in all modes.

I will soon replace which-key labels with docstrings to get around this, but until then we're stuck without which-key support.

File: modules/lang/ruby/config.el [hunk 2]
```diff
@@ -58,7 +58,8 @@
         "'"  #'robe-start
         ;; robe mode specific
         "h"  #'robe-doc
-        "rr" #'robe-rails-refresh
+        :prefix "r"
+        "r" #'robe-rails-refresh
```
Comment (hlissner) [inline modules/lang/ruby/config.el]: The `;prefix "r"` inadvertantly applies to the keybinds below it. Please revert this change.

File: modules/lang/ruby/config.el (hunk 1 above)
Comment (alexandergraul) [reply modules/lang/ruby/config.el]: I see, the lables are one of the reasons I labled it WIP as I as not happy with it anyway. I'll remove them.

File: modules/lang/ruby/config.el (hunk 2 above)
Comment (alexandergraul) [reply modules/lang/ruby/config.el]: I thought the `:prefix "s"` would override my `:prefix "r"`a for the keybindings below it. I don't mind removing it, just thought it was cleaner to write the prefix explicit instead of "rr"

Comment (hlissner) [reply modules/lang/ruby/config.el]: Prefixes stack. `:prefix "a" :prefix "b" "c" #'do-something` binds to the `"a b c"` key sequence.

You'll have to enclose them to isolate them: `(:prefix "a" ...) (:prefix "b" ...)`

File: modules/lang/ruby/config.el [hunk 3]
```diff
@@ -163,3 +163,58 @@
...
+  :hook (enh-ruby-mode . projectile-rails-mode)
+  :config
+  (map! :localleader
+        :map projectile-rails-mode-map
+        ;; projectile-rails also has a projectile-rails-command-map
+        ;; There are two reasons for manual keybindings instead of mapping
+        ;; prefix "r" to it:
+        ;; - prefix "r" is also used by robe
+        ;; - Some of the keybindings differ
+        (:prefix ("r" . "Rails")
```
Comment (hlissner) [inline modules/lang/ruby/config.el]: The which-key labels have to be removed from :prefix as well:

```diff
-(:prefix ("r" . "Rails")
+(:prefix "r"
```

File: modules/lang/ruby/config.el [hunk 4]
```diff
@@ -163,3 +163,58 @@
...
+        "b" #'projectile-rails-find-job
+        "x" #'projectile-rails-extract-region
+        ;; Go to files
+        "gg" #'projectile-rails-goto-gemfile
+        "gr" #'projectile-rails-goto-routes
+        "gd" #'projectile-rails-goto-schema
+        "gs" #'projectile-rails-goto-seeds
+        "gh" #'projectile-rails-goto-spec-helper)
+        ;; Open consoles etc.
+        (:prefix ("o" . "open")
```
Comment (hlissner) [inline modules/lang/ruby/config.el]: Same as above

**Comment by otavioschwanck** [comment]
Very nice work, love the doom emacs community <3

File: modules/lang/ruby/config.el (hunk 3 above)
Comment (otavioschwanck) [inline modules/lang/ruby/config.el]: This prefix will work on entire project or only on enh-ruby-mode ?

In spacemacs, you only can use projectile-rails shortcuts on enh-ruby-mode files, and it is very bad.  On pure emacs with projectile-rails, you can use the projectile-rails shortcuts on any mode (any file).

**Comment by otavioschwanck** [comment]
image

When search for projectile-commands with M-x, the shortcuts appears wrong

**Comment by otavioschwanck** [comment]
image

Ruby on Rails have many type of files on your workflow.  html, css, JS, .rb, .arb, .yml.

Its super necessary to have access to projectile-rails on those files too.  The projectile-rails have commands to search JS, CSS, etc.  

I tried here, the SPC m r * only works on .rb files.  Would be very nice enable projectile-rails commands on whole project.  Or at least, the possibility to do this.

File: modules/lang/ruby/config.el (hunk 2 above)
Comment (alexandergraul) [reply modules/lang/ruby/config.el]: ah, good to know!

**Comment by alexandergraul** [comment]
> When search for projectile-commands with M-x, the shortcuts appears wrong

For me they show up correctly , but I've also just removed those prefix descriptions.

> Ruby on Rails have many type of files on your workflow. html, css, JS, .rb, .arb, .yml.

That is true, I'll see what modes to hook it into additionally

**Comment by otavioschwanck** [comment]
> > When search for projectile-commands with M-x, the shortcuts appears wrong
> 
> For me they show up correctly , but I've also just removed those prefix descriptions.
> 
> > Ruby on Rails have many type of files on your workflow. html, css, JS, .rb, .arb, .yml.
> 
> That is true, I'll see what modes to hook it into additionally

Some modes that i use on rails project:

coffee-mode, json-mode, yaml-mode, js2-mode, css-mode, scss-mode, javascript-mode, projectile-rails-server-mode, inf-ruby-mode, rspec-mode

**Comment by alexandergraul** [comment]
@otavioschwanck I've added a hook into web-mode (if you enable the :lang web module). That way you will have it at least in .html and .css files. For yaml I am not sure if it is good idea to hook it into, yaml is used for a lot more than just rails. You can always do it manually `(add-hook 'yaml-mode-hook #'projectile-rails-mode)`

js2-mode already uses `<localleader> r`, I don't know what to do about that :(

**Comment by otavioschwanck** [comment]
> @otavioschwanck I've added a hook into web-mode (if you enable the :lang web module). That way you will have it at least in .html and .css files. For yaml I am not sure if it is good idea to hook it into, yaml is used for a lot more than just rails. You can always do it manually `(add-hook 'yaml-mode-hook #'projectile-rails-mode)`
> 
> js2-mode already uses `<localleader> r`, I don't know what to do about that :(

maybe we can skip js-mode for this, sadly =\.  Good work anyway.

**Comment by otavioschwanck** [comment]
I forget one mode, magit-status-mode.  Many times i use the SPC m r * on this.

**Comment by otavioschwanck** [comment]
I found something strange:
image

Opened two .rb files, one gets enhruby and the other enhruby 2.6.3.  The SPC m r * only works on enhruby 2.6.3.

i'm using +rbenv.

**Comment by otavioschwanck** [comment]
Another problem:

The Rails Server and Console are not listed on SPC , (buffer-list).

If you run the server, close, and try to open again (SPC m o s), it trys to open another server instead opening the buffer of last created server (like projectile-rails on vanilla emacs). it gives a error (port already in use)

File: modules/lang/ruby/config.el [hunk 5]
```diff
@@ -163,3 +163,61 @@
...
+        "gr" #'projectile-rails-goto-routes
+        "gd" #'projectile-rails-goto-schema
+        "gs" #'projectile-rails-goto-seeds
+        "gh" #'projectile-rails-goto-spec-helper)
+        ;; Open consoles etc.
+        (:prefix "o"
+        "d" #'projectile-rails-dbconsole
+        "c" #'projectile-rails-console
+        "s" #'projectile-rails-server
+        "g" #'projectile-rails-generate)))
```
Comment (hlissner) [inline modules/lang/ruby/config.el]: Please correct the indentation of these two :prefix blocks (by one space).

**Comment by hlissner** [comment]
> Another problem:
> 
> The Rails Server and Console are not listed on SPC , (buffer-list).

This is because Doom doesn't register them as "real" buffers. Add `doom-mark-buffer-as-real-h` to their mode's hooks.

File: modules/lang/ruby/config.el (hunk 5 above)
Comment (otavioschwanck) [reply modules/lang/ruby/config.el]: would be nice rename the projectile-rails to something else:

image


its impossible to see what the commands do.

Maybe change it to rails ?

Comment (hlissner) [reply modules/lang/ruby/config.el]: That's not really up to us. Perhaps `(setq which-key-show-docstrings 'docstring-only)` may help?

**Comment by alexandergraul** [comment]
If you want projectile-rails to always be available you can also call `(projectile-rails-global-mode)` in your private config.

File: modules/lang/ruby/config.el (hunk 5 above)
Comment (luisenrike) [reply modules/lang/ruby/config.el]: There's a way to remove `projectile-rails` 
https://github.com/justbur/emacs-which-key#key-and-description-replacement

```
(push '((nil . "projectile-rails-\\(.+\\)") . (nil . "\\1"))
        which-key-replacement-alist)
```

**Comment by luisenrike** [comment]
Some of commands are missing, this is one of the cons of having to manually map them, I would just assign `projectile-rails-command-map` to a different prefix and just use the default keybindings. (see [this](https://github.com/hlissner/doom-emacs/pull/2463#discussion_r374496846) for how to remove the `projectile-rails` prefix)

```
projectile-rails-find-component
projectile-rails-find-current-javascript
projectile-rails-find-current-serializer
projectile-rails-find-current-stylesheet
projectile-rails-find-rake-task
projectile-rails-find-serializer
projectile-rails-find-webpack
projectile-rails-goto-file-at-point
projectile-rails-mode-goto-map
projectile-rails-mode-run-map
```

**Comment by alexandergraul** [comment]
Thanks for double checking! Yes, it seems that doing it manually is not a good way if I already missed a few. I'd really like to stick to "r" as the prefix, I can't find any other that is as intuitive to use. robe only uses `<localleader> r r` for `robe-rails-refresh`. While this is a breaking change, maybe we can change that single keybinding and keep "r" for `projectile-rails-command-map`?

**Comment by luisenrike** [comment]
Yeah, `r` is the best option for `projectile-rails`, maybe move robe to `R`?

**Comment by alexandergraul** [comment]
That might work, we can lose the prefix as well for a single command then

**Comment by otavioschwanck** [comment]
Sometimes, the rails-server (SPC m o s) closes when press q, instead of only hiding it.

**Comment by otavioschwanck** [comment]
> I found something strange:
> image
> 
> Opened two .rb files, one gets enhruby and the other enhruby 2.6.3. The SPC m r * only works on enhruby 2.6.3.
> 
> i'm using +rbenv.
> 
> It occurs on the first file opened with projectile.

About this, it is related to +rbenv flag.  Maybe the +rbenv change the mode and this PR changes doen't work.



> Sometimes, the rails-server (SPC m o s) closes when press q, instead of only hiding it.
> 
> Keep pressing SPC m o s, then, q. On second tries, it close and start a new server.

Also related to +rbenv flag.  Dont know why.

**Comment by otavioschwanck** [comment]
Since this PR improves ruby support, add a shortcut to robe-jump.  Its a very nice feature.

File: modules/lang/ruby/config.el (hunk 5 above)
Comment (alexandergraul) [reply modules/lang/ruby/config.el]: I've included this change in  97ee4b4

**Comment by alexandergraul** [comment]
@otavioschwanck robe-jump is already on `<leader> c d` (`+lookup/definition`)

**Comment by alexandergraul** [comment]
@otavioschwanck  I am not sure I understand the issue - are you saying that the rails console and rails server processes are restarted sometimes? Is this only a problem when using them with Doom (this PR) or does it also happen with projectile-rails without Doom?

**Comment by otavioschwanck** [comment]
> @otavioschwanck I am not sure I understand the issue - are you saying that the rails console and rails server processes are restarted sometimes? Is this only a problem when using them with Doom (this PR) or does it also happen with projectile-rails without Doom?

it only occurs on Doom.  I've my .emacs: https://github.com/otavioschwanck/emacs_on_rails and it never happened.

The expected behaviour is after start console \ server and try to open again, it just show the buffer instead starting a new server \ console.

**Comment by luisenrike** [comment]
Hi @otavioschwanck, does it happen only with this PR and not with the latest doom update? If not that's a different issue (seems like it) and not related to this one. It would be good to know this so @alexandergraul can get this one merged 🤞 , really need this stuff

**Comment by otavioschwanck** [comment]
> Hi @otavioschwanck, does it happen only with this PR and not with the latest doom update? If not that's a different issue (seems like it) and not related to this one. It would be good to know this so @alexandergraul can get this one merged  , really need this stuff

i've my own config with projectile-rails with doom (installed manually) and it happens too.  I dont know  why it happen.  Without server and console, projectile-rails lose so much productivity.

**Comment by alexandergraul** [comment]
For me personally, server and console are nice but not the main reason I'm using projectile-rails (which is navigation). I think we should create an issue for the closing of the console/server processes and investigate why that happens after this PR is merged. I don't mind living on this branch for my personal use, but others might want to use projectile-rails without setting it up themselves.

What do you think @hlissner ?

**Comment by hlissner** [comment]
@otavioschwanck wrt the missing `SPC m r` prefix, does this fix your issue?

```elisp
(add-hook 'projectile-rails-mode-hook #'evil-normalize-keymaps)
```

As for the missing rails issue, this is because the popup manager auto-kills buffers (and their associated processes) after you close the popup by default. A popup rule needs to defined for the console buffer. e.g.

```elisp
(after! projectile-rails
  (set-popup-rule! REGEXP :ttl nil))
```

You must replace `REGEXP` with a regex that matches the name of the rails console buffer (open it and run `M-: (buffer-name)` in it).

**Comment by otavioschwanck** [comment]
> @otavioschwanck wrt the missing `SPC m r` prefix, does this fix your issue?
> 
> ```emacs-lisp
> (add-hook 'projectile-rails-mode-hook #'evil-normalize-keymaps)
> ```
> 
> As for the missing rails issue, this is because the popup manager auto-kills buffers (and their associated processes) after you close the popup by default. A popup rule needs to defined for the console buffer. e.g.
> 
> ```emacs-lisp
> (after! projectile-rails
>   (set-popup-rule! REGEXP :ttl nil))
> ```
> 
> You must replace `REGEXP` with a regex that matches the name of the rails console buffer (open it and run `M-: (buffer-name)` in it).

i think it need be added in this PR. @alexandergraul

File: modules/lang/ruby/config.el [hunk 6]
```diff
@@ -163,3 +163,17 @@
...
+
+
+(use-package! projectile-rails
+  :when (featurep! +rails)
+  :hook (enh-ruby-mode . projectile-rails-mode)
+  :init
+  (when (featurep! :lang web)
+    (add-hook 'web-mode-hook #'projectile-rails-mode))
+  :config
+  (map! :localleader
```
Comment (otavioschwanck) [inline modules/lang/ruby/config.el]: (add-hook 'projectile-rails-mode-hook #'evil-normalize-keymaps)

Comment (otavioschwanck) [inline modules/lang/ruby/config.el]: It works, but maybe have a better way (with regex):

  (after! projectile-rails
    (set-popup-rule! "*rails*" :ttl nil)
    (set-popup-rule! "*projectile-rails-server*" :ttl nil))

**Comment by otavioschwanck** [comment]
I added in comments the fixes for the two bugs that i listed.  Thanks @hlissner !

File: modules/lang/ruby/config.el (hunk 6 above)
Comment (alexandergraul) [reply modules/lang/ruby/config.el]: I've added the hook to the `:hook` section, afaik that should do the same

**Comment by alexandergraul** [comment]
@hlissner I've added the two fixes, can you take another look please?

File: modules/lang/ruby/config.el (hunk 6 above)
Comment (otavioschwanck) [reply modules/lang/ruby/config.el]: nice

**Comment by otavioschwanck** [comment]
> @hlissner I've added the two fixes, can you take another look please?

already tested here, working flayless.  Good job @alexandergraul !

image

One idea:

On default init.el, change from:

;; ruby

to

;; (ruby
;;  +rails)

99% of ruby programmers use Rails

File: modules/lang/ruby/config.el (hunk 6 above)
Comment (hlissner) [reply modules/lang/ruby/config.el]: Please remove `evil-normalize-keymaps` from `:hook` and move it into an `(add-hook 'projectile-rails-mode-hook #'evil-normalize-keymaps)` under :config.

`use-package` has a side-effect where it associates :hook functions with the containing package, as if they belonged to them. i.e. `(autoload 'evil-normalize-keymaps "projectile-rails")`, which can cause errors when Emacs tries to load `projectile-rails` looking for `evil-normalize-keymaps`. 

Since, in this particular case, evil is always eager loaded at startup, this isn't a problem here, but it will come back to bite us if a) `:lang (ruby +rails)` were loaded before `:editor evil` (for whatever reason) or b) we one day start lazy loading evil.

Comment (hlissner) [reply modules/lang/ruby/config.el]: Also: the evil-normalize-keymaps call should be conditional, otherwise it will throw errors for non-evil users.

**Comment by alexandergraul** [comment]
Thanks for the info about `use-package`. I also forgot about non-evil users :see_no_evil: 

I've done the requested changes.

**Comment by otavioschwanck** [comment]
Tested the changes, still works!

**Comment by hlissner** [comment]
Looks good to me. Thanks for all your help!
